### PR TITLE
Add test tools to make it easier to test Network/reauthentication logic

### PR DESCRIPTION
**Pull request by marcaaron**
cc @marcochavezf 

### Details

**Test Tools**

A few testing tools are being added here so we can test conditions

- Invalid credentials
- Invalid authToken
- All requests are failing when run (e.g. auth down scenario)

### Fixed Issues

https://github.com/Expensify/Expensify/issues/207534

### Tests
1. Build the app in staging mode

**Note:** for native platforms this diff needed to be applied to get "staging"

```diff
diff --git a/src/libs/Environment/getEnvironment/index.native.js b/src/libs/Environment/getEnvironment/index.native.js
index 89547e0ee..1cfd0e0dd 100644
--- a/src/libs/Environment/getEnvironment/index.native.js
+++ b/src/libs/Environment/getEnvironment/index.native.js
@@ -13,8 +13,8 @@ let environment = null;
 function getEnvironment() {
     return new Promise((resolve) => {
         // If we've already set the environment, use the current value
-        if (environment) {
-            return resolve(environment);
+        if (true) {
+            return resolve(CONST.ENVIRONMENT.STAGING);
         }

         if (lodashGet(Config, 'ENVIRONMENT', CONST.ENVIRONMENT.DEV) === CONST.ENVIRONMENT.DEV) {
```

1. Navigate to Preferences and verify there are new testing preferences available
1. Toggle on "Simulate failing network requests"

1. Proceed to do something in the app that triggers a network request (navigate to chat)
1. Verify requests are shown as retrying in the logs and eventually fail 

1. Go back to preferences and untoggle this option
1. Invalidate the authToken via the button that says "Invalidate"

1. Do something in the app that triggers a network request
1. Verify that a call to `Authenticate` appears in the Chrome Dev Tools network tab and you are reauthenticated

1.  Go back to preferences and invalidate the authToken and destroy the device credentials

1. Do something in the app that triggers a network request
1. Verify you are logged out 
 
- [ ] Verify that no errors appear in the JS console

### PR Review Checklist
<details>
<summary><h4>PR Reviewer Checklist</h4></summary>

- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

</details>

### QA Steps
1. Navigate to Preferences and verify there are new testing preferences available
1. Toggle on "Simulate failing network requests"
1. Proceed to do something in the app that will trigger a network request
1. Verify requests are shown as retrying in the logs and eventually fail  
1. Go back to preferences and untoggle the "Simulate failing network requests" option
1. Invalidate the authToken via the button that says "Invalidate"
1. Do something in the app that triggers a network request
1. Verify that a call to `Authenticate` appears in the Chrome Dev Tools network tab and you are reauthenticated (Web Only)
1.  Go back to preferences and invalidate the authToken and press the "Destroy" button to corrupt the device credentials
1. Do something in the app that triggers a network request
1. Verify you are logged out 

- [ ] Verify that no errors appear in the JS console

### CODE REVIEW

File: src/libs/Log.js [hunk 1]
```diff
@@ -37,6 +37,7 @@ function LogCommand(parameters) {
 function serverLoggingCallback(logger, params) {
     const requestParams = params;
     requestParams.shouldProcessImmediately = false;
+    requestParams.shouldRetry = false;
```
Comment (marcaaron) [inline src/libs/Log.js:40]: One weird side effect of trigger the option where all the requests should fail is that some iteration of a Log command is getting retried forever. I suspect it's something like:

1. Log fails
2. We run out of retries for a Log command
3. We log that we ran out of retries and log again
4. etc.

But not 100% sure and need to debug some more. I think `Log` is a "persistable" request (but isn't today) so once we have a backoff logic in place we can retry it forever - but at some rate of backoff. For now, we would be making many requests over and over again if we had a situation like "Auth down" so I am making Log not retryable.

**Comment by marcaaron** [comment]
Still testing on some platforms but taking out of draft.

File: src/libs/Log.js (hunk 1 above)
Comment (Justicea83) [reply src/libs/Log.js:40]: Yeah, I think we need more debugging to find out that

**Review by Justicea83 — COMMENTED** [review]
Looks good overall.

File: src/libs/HttpUtils.js [hunk 2]
```diff
@@ -11,6 +11,12 @@ Onyx.connect({
     callback: val => shouldUseSecureStaging = (val && _.isBoolean(val.shouldUseSecureStaging)) ? val.shouldUseSecureStaging : false,
 });
 
+let shouldFailAllRequests = false;
+Onyx.connect({
+    key: ONYXKEYS.NETWORK,
+    callback: val => shouldFailAllRequests = (val && _.isBoolean(val.shouldFailAllRequests)) ? val.shouldFailAllRequests : false,
```
[the remark this answers is not in the document]
Comment (marcaaron) [reply src/libs/HttpUtils.js:17]: What would you suggest?

File: src/components/TestToolMenu.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    <>
+        <Text style={[styles.formLabel]} numberOfLines={1}>
+            Test Preferences
+        </Text>
+
+        {/* Option to switch from using the staging secure endpoint or the production secure endpoint.
+        This enables QA and internal testers to take advantage of sandbox environments for 3rd party services like Plaid and Onfido. */}
+        <TestToolRow title="Use Secure Staging Server">
+            <Switch
+                isOn={props.user.shouldUseSecureStaging || false}
```
[the remark this answers is not in the document]
Comment (marcaaron) [reply src/components/TestToolMenu.js:48]: agree!

File: src/libs/HttpUtils.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (Justicea83) [reply src/libs/HttpUtils.js:17]: Here is a simple illustration ,NAB

Screenshot 2022-04-22 at 9 16 24 PM

**Review by Justicea83 — DISMISSED** [review]
LGTM!

**Comment by Justicea83** [comment]
all you @marcaaron

File: src/components/TestToolMenu.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    <>
+        <Text style={[styles.formLabel]} numberOfLines={1}>
+            Test Preferences
+        </Text>
+
+        {/* Option to switch from using the staging secure endpoint or the production secure endpoint.
+        This enables QA and internal testers to take advantage of sandbox environments for 3rd party services like Plaid and Onfido. */}
+        <TestToolRow title="Use Secure Staging Server">
+            <Switch
+                isOn={props.user.shouldUseSecureStaging}
```
Comment (marcochavezf) [inline src/components/TestToolMenu.js:48]: When I tested locally, I got this console warning:

Screen Shot 2022-04-22 at 16 34 55

I think we can fix that by adding the default values to `Onyx.init`, no?

https://github.com/Expensify/App/blob/5b8b75cca48740db41a6af50452c9312e3e80510/src/setup/index.js#L26-L36

Comment (marcaaron) [reply src/components/TestToolMenu.js:48]: Oh hmm I guess if there is a `user` object then it won't use the defaults. Sorry @Justicea83 I am going to put the short circuit back in here. @marcochavezf not sure if this is something for `initialKeyStates`- the code was fine before so I'm going to revert that last change.

File: src/libs/HttpUtils.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (marcaaron) [reply src/libs/HttpUtils.js:17]: Hmm sorry, I don't understand the suggestion.

File: src/components/TestToolMenu.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (marcaaron) [reply src/components/TestToolMenu.js:48]: Sorry this was actually not correct - resolving :D

File: src/components/TestToolMenu.js (hunk 4 above)
Comment (Justicea83) [reply src/components/TestToolMenu.js]: Sure that's fine, we look good now!

**Review by Justicea83 — APPROVED** [review]
All you @marcaaron

**Review by marcochavezf — APPROVED** [review]
LGTM, this test tool is cool!

File: src/libs/HttpUtils.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (kidroca) [reply src/libs/HttpUtils.js:17]: These should do: 

```js
    callback: val => shouldFailAllRequests = lodashGet(val, 'shouldFailAllRequests', false);
```

or 

```js
callback: val => shouldFailAllRequests = Boolean(val && val.shouldFailAllRequests);
```

[the remark this answers is not in the document]
Comment (marcaaron) [reply src/libs/HttpUtils.js:17]: Oh hmm but why? Anything prompt you to leave this comment @kidroca? I'm not seeing an issue or concern. Let me know, thanks!

[the remark this answers is not in the document]
Comment (kidroca) [reply src/libs/HttpUtils.js:17]: No issue, this PR was shared on slack, decided to take a look and saw this comment

> I know this is copy paste, but can't it be simplified?

A ternary that resolves to a boolean value can be replaced by a boolean expression for cleaner code

**Pull request merged**